server/src/main/java/com/tencent/rss/server/buffer/ShuffleBuffer.java

```diff
@@ -0,0 +1,374 @@
+            if (eventIndex < sortedEventId.size() - 1) {
+              fromEventId = sortedEventId.get(eventIndex + 1);
+            }
+            hasBlockId = true;
+            // read from 1st block with following inFlushBlock/blocks
+            currentBlockId = Constants.INVALID_BLOCK_ID;
+            updatedLastBlockId = Constants.INVALID_BLOCK_ID;
+          }
+        }
+        if (offset >= readBufferSize) {
+          break;
+        }
+        eventIndex++;
+      }
+    }
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      ListReadInfo listReadInfo = processCachedBlocks(
+          offset, blocks, currentBlockId, readBufferSize, bufferSegments);
+      offset = listReadInfo.offset;
+      hasBlockId = listReadInfo.hasBlockId;
+    }
+    if (lastBlockId != Constants.INVALID_BLOCK_ID && offset == 0 && !hasBlockId) {
+      // can't find lastBlockId, it should be flushed
+      // try read with blockId = Constants.INVALID_BLOCK_ID
+      return updateBufferSegments(Constants.INVALID_BLOCK_ID, readBufferSize, bufferSegments);
+    } else {
+      return new ActualReadInfo(offset, fromEventId, updatedLastBlockId);
+    }
+  }
+
+  private void updateShuffleData(
+      long lastBlockId,
+      long readBufferSize,
+      byte[] data,
+      long fromEventId) throws Exception {
+    List<Long> sortedEventId = sortFlushingEventId();
+    long currentBlockId = lastBlockId;
+    int offset = 0;
+    if (fromEventId != INVALID_EVENT_ID) {
+      // read data start with inFlushBlock
+      for (Long eventId : sortedEventId) {
+        if (eventId >= fromEventId) {
+          ListReadInfo listReadInfo = processCachedBlocks(offset, inFlushBlockMap.get(eventId),
+              currentBlockId, readBufferSize, data);
+          offset = listReadInfo.offset;
+          // from next flush/cache buffer, read from 1st block
+          currentBlockId = Constants.INVALID_BLOCK_ID;
+          if (offset >= readBufferSize) {
+            break;
+          }
+        }
+      }
+    }
+    if (blocks.size() > 0 && offset < readBufferSize) {
+      processCachedBlocks(offset, blocks, currentBlockId, readBufferSize, data);
+    }
+  }
+
+  private List<Long> sortFlushingEventId() {
+    List<Long> eventIdList = Lists.newArrayList(inFlushBlockMap.keySet());
+    eventIdList.sort((id1, id2) -> {
+      if (id1 > id2) {
+        return 1;
+      }
+      return -1;
+    });
+    return eventIdList;
+  }
+
+  private ListReadInfo processCachedBlocks(
+      int offset,
+      List<ShufflePartitionedBlock> cachedBlocks,
+      long lastBlockId,
+      long readBufferSize,
+      Object parameter) throws Exception {
+    int currentOffset = offset;
+    Callable<Integer> callable = null;
+    if (lastBlockId == Constants.INVALID_BLOCK_ID) {
+      // read from first block
+      for (ShufflePartitionedBlock block : cachedBlocks) {
+        // do action
+        if (parameter instanceof byte[]) {
```

Comment:
It may be possible and appropirate to use visitor pattern and double dispatch to avoid these instanceof matching.

Reply:
it will be removed for better readable

Reply:
> It may be possible and appropirate to use visitor pattern and double dispatch to avoid these instanceof matching.

+1, Could we extract the way that visit block.

Reply:
Readable is important, will remove such trick code